### PR TITLE
boost182: init at 1.82.0

### PR DESCRIPTION
**Pull request by afh**
###### Description of changes

ℹ️ Boost 1.82 brings a fix for Boost Python relevant for Python 3.11 (see https://github.com/boostorg/python/pull/385)

❓ Please let me know if this PRs merge-base should rather be `staging` than `master`…

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
Build is failing on ofborg: https://github.com/NixOS/nixpkgs/pull/227057/checks?check_run_id=12864525140, please investigate.

**Comment by afh** [comment]
Here's what I've found, @NickCao:

* The run succeeds on darwin (both x86_64 and aarch64) but fails on linux (both x86_64 and aarch64).
* The cause for the failure on linux is: `./bootstrap.sh: line 13: /build/boost_1_82_0/tools/build/src/engine/build.sh: not found`
* Inspecting `boostrap.sh` `$my_dir/tools/build/src/engine/build.sh` is called on [on line 229](https://github.com/boostorg/boost/blob/master/bootstrap.sh#L229)
* `my_dir` is set using `my_dir=$(dirname $0)` on [line 195](https://github.com/boostorg/boost/blob/master/bootstrap.sh#L195)
* The last change to `bootstrap.sh` was 2 years ago
* Inspecting `$src` from `nix shell nixpkgs-dev#boost182` where `nixpkgs-dev` points to a local git repository with this PRs branch checked out, the `tools/build/src/engine/build.sh` file exists, i.e.:
```
% tar jtf $src | grep engine/build.sh
boost_1_82_0/tools/build/src/engine/build.sh
```

❓ Could this be an issue related to the test infrastructure and OfBorg rather than an issue with the boost derivation and its source?

**Comment by NickCao** [comment]
It's due to https://github.com/boostorg/build/commit/2bce4ad599468790abeed547cd0b7bc52ca69993, we need to patchShebangs then.

File: pkgs/development/libraries/boost/generic.nix [hunk 1]
```diff
@@ -174,6 +174,10 @@ stdenv.mkDerivation {
     inherit boostBuildPatches;
   };
 
+  postPatch = ''
+    patchShebangs tools/build/src/engine/build.sh
+  '';
```
Comment (wegank) [inline pkgs/development/libraries/boost/generic.nix:179]: This should only apply when `lib.versionAtLeast version "1.82"` is true.

Comment (afh) [reply pkgs/development/libraries/boost/generic.nix]: I was wondering whether that would be necessary or not, now I think I understand that without the version check there are a lot of rebuilds, correct?

Comment (wegank) [reply pkgs/development/libraries/boost/generic.nix]: Yes. In the meantime, I think that `postPatch = ""` will still cause rebuilds, so maybe `postPatch = if cond then expr else null` instead.

Comment (afh) [reply pkgs/development/libraries/boost/generic.nix]: Thanks for the helpful info, @wegank. I've verified that on my machine `boost181` isn't rebuild. 👍

**Comment by NickCao** [comment]
Build still failing for the same reason....

**Comment by afh** [comment]
That makes sense, @NickCao, the suggestion to `patchShebangs`—while helpful—changes the `tools/build/src/engine/build.sh` script, but not `bootstrap.sh`, in which the error occurs and which calls the `…engine/build.sh`.

I'm going over the source code while reading through the logs. What I cannot explain is why in [adf](https://logs.ofborg.org/?attempt_id=eaf317e3-ec7a-4bdd-aca7-e3232b986c1b&key=nixos%2Fnixpkgs.227057) the `/nix/store/ssy83226flb80lkhnynkky3sz7hfjfwp-boost-1.79.0.drv` derivation is also built.

Is there a way to run `bootstrap.sh` on ofborg with `-x` ? I see that `boost/generic.nix` sets it as the `configureScript` what possibly implications come with that in terms of the value of `$0`?

**Comment by NickCao** [comment]
Turns out that we were looking at the wrong place:
```diff
--- a/pkgs/development/tools/boost-build/default.nix
+++ b/pkgs/development/tools/boost-build/default.nix
@@ -14,7 +14,7 @@ let
   defaultVersion = "4.4.1";
 in

-stdenv.mkDerivation {
+stdenv.mkDerivation rec {
   pname = "boost-build";
   version =
     if useBoost ? version
@@ -39,6 +39,8 @@ stdenv.mkDerivation {
   postPatch = ''
     substituteInPlace src/build-system.jam \
     --replace "default-toolset = darwin" "default-toolset = clang-darwin"
+  '' + lib.optionalString (lib.versionAtLeast version "1.82") ''
+    patchShebangs --build src/engine/build.sh
   '';

   nativeBuildInputs = [
```

**Comment by afh** [comment]
Great find, @NickCao!! That also explains the `substituteStream(): WARNING: pattern 'default-toolset = darwin' doesn't match anything in file 'src/build-system.jam'` in the OfBorg logs. I didn't know about the boost-build derivation.

Should I change the `default-toolset` substitute only applied when `lib.optionalString stdenv.isDarwin` while I'm at it?

**Comment by afh** [comment]
And is the patchShebangs in `boost/generic.nix` still necessary? I'd think so, but want to make sure the changes to bring Boost 1.82.0 to nixpkgs are minimal.

**Comment by afh** [comment]
Alright I've found the actual script that fails, it is [`bootstrap.sh`](https://github.com/boostorg/build/blob/9f488e003a568dffe0caed05d86ed6f1a8f8c7f3/bootstrap.sh#L13) in the `build/` subdirectory/module not the repo top-level one.

Could this be an issue with submodule source not (yet?) having fully checked out?

File: pkgs/development/tools/boost-build/default.nix [hunk 2]
```diff
@@ -14,7 +14,7 @@ let
   defaultVersion = "4.4.1";
 in
 
-stdenv.mkDerivation {
+stdenv.mkDerivation rec {
```
Comment (wegank) [inline pkgs/development/tools/boost-build/default.nix:17]:
```suggestion
stdenv.mkDerivation {
```

File: pkgs/development/tools/boost-build/default.nix [hunk 3]
```diff
@@ -39,6 +39,8 @@ stdenv.mkDerivation {
   postPatch = ''
     substituteInPlace src/build-system.jam \
     --replace "default-toolset = darwin" "default-toolset = clang-darwin"
+  '' + lib.optionalString (lib.versionAtLeast version "1.82") ''
```
Comment (wegank) [inline pkgs/development/tools/boost-build/default.nix:42]:
```suggestion
  '' + lib.optionalString (useBoost ? version && lib.versionAtLeast useBoost.version "1.82") ''
```

Comment (afh) [reply pkgs/development/tools/boost-build/default.nix:42]: Nicely done! Thanks for the suggestion!!

File: pkgs/development/libraries/boost/generic.nix [hunk 4]
```diff
@@ -174,6 +174,12 @@ stdenv.mkDerivation {
     inherit boostBuildPatches;
   };
 
+  postPatch =
+    if lib.versionAtLeast version "1.82" then
+      "patchShebangs tools/build/src/engine/build.sh"
+    else
+      null;
```
Comment (wegank) [inline pkgs/development/libraries/boost/generic.nix:181]:
```suggestion
```

File: pkgs/development/tools/boost-build/default.nix [hunk 5]
```diff
@@ -39,6 +39,8 @@ stdenv.mkDerivation {
   postPatch = ''
```
Comment (afh) [inline pkgs/development/tools/boost-build/default.nix:39]: @wegank What are your thoughts on this?

```suggestion
  postPatch = lib.optionalString stdenv.isDarwin ''
```

Comment (wegank) [reply pkgs/development/tools/boost-build/default.nix:39]: I prefer to keep patches as general as possible, so I'm not encouraging this.

Comment (afh) [reply pkgs/development/tools/boost-build/default.nix:39]: Thanks for the helpful clarification 🙂

**Comment by wegank** [comment]
@ofborg build boost182

**Comment by afh** [comment]
Kindly re-requesting review, now that all tests have passed. Thanks for all the helpful comments to make this PR the best it can be, @wegank, very much appreciated! 🙏

**Review by wegank — APPROVED** [review]

**Comment by wegank** [comment]
Please squash the commits, thanks!

**Comment by afh** [comment]
Done! ✅🙂👍

**Pull request merged**